From ffbe29b9305ecaf203ad550eca2e154bbdbd8c90 Mon Sep 17 00:00:00 2001
From: Dominique Martinet <asmadeus@codewreck.org>
Date: Sun, 4 Oct 2020 17:26:54 +0200
Subject: [PATCH] doc: warn that preallocation disables compression on btrfs

falloc() will disable cow for writes in the preallocated area, and
has the side-effect of also disabling file compression for that file.

If user wants compression they should be advised to use trunc or none
file-allocation instead.
---
 doc/manual-src/en/aria2c.rst |  7 ++++---
 doc/manual-src/pt/aria2c.rst |  7 ++++---
 src/usage_text.h             | 13 ++++++++-----
 3 files changed, 16 insertions(+), 11 deletions(-)

diff --git a/doc/manual-src/en/aria2c.rst b/doc/manual-src/en/aria2c.rst
index f093f7335..41e2a9f06 100644
--- a/doc/manual-src/en/aria2c.rst
+++ b/doc/manual-src/en/aria2c.rst
@@ -1357,9 +1357,10 @@ Advanced Options
   If you are using newer file systems such as ext4
   (with extents support), btrfs, xfs or NTFS(MinGW build only), ``falloc`` is
   your best choice. It allocates large(few GiB)
-  files almost instantly. Don't use ``falloc`` with
-  legacy file systems such as ext3 and FAT32 because it takes
-  almost same time as ``prealloc`` and it blocks aria2
+  files almost instantly. Note that btrfs disables compression when
+  pre-allocating, use ``trunc`` or ``none`` if you want compression.
+  Don't use ``falloc`` with legacy file systems such as ext3 and FAT32
+  because it takes almost same time as ``prealloc`` and it blocks aria2
   entirely until allocation finishes. ``falloc`` may
   not be available if your system doesn't have
   :manpage:`posix_fallocate(3)` function.
diff --git a/doc/manual-src/pt/aria2c.rst b/doc/manual-src/pt/aria2c.rst
index 668a3331a..5266118b3 100644
--- a/doc/manual-src/pt/aria2c.rst
+++ b/doc/manual-src/pt/aria2c.rst
@@ -1204,9 +1204,10 @@ Opções Avançadas
   If you are using newer file systems such as ext4
   (with extents support), btrfs, xfs or NTFS(MinGW build only), ``falloc`` is
   your best choice. It allocates large(few GiB)
-  files almost instantly. Don't use ``falloc`` with
-  legacy file systems such as ext3 and FAT32 because it takes
-  almost same time as ``prealloc`` and it blocks aria2
+  files almost instantly. Note that btrfs disables compression when
+  pre-allocating, use ``trunc`` or ``none`` if you want compression.
+  Don't use ``falloc`` with legacy file systems such as ext3 and FAT32
+  because it takes almost same time as ``prealloc`` and it blocks aria2
   entirely until allocation finishes. ``falloc`` may
   not be available if your system doesn't have
   :manpage:`posix_fallocate(3)` function.
diff --git a/src/usage_text.h b/src/usage_text.h
index ebf67e3d3..d3fe4be20 100644
--- a/src/usage_text.h
+++ b/src/usage_text.h
@@ -138,12 +138,15 @@
     "                              the file.\n"                         \
     "                              If you are using newer file systems such as ext4\n" \
     "                              (with extents support), btrfs, xfs or NTFS\n" \
-    "                              (MinGW build only), 'falloc' is your best\n"   \
+    "                              (MinGW build only), 'falloc' is your best\n" \
     "                              choice. It allocates large(few GiB) files\n" \
-    "                              almost instantly. Don't use 'falloc' with legacy\n" \
-    "                              file systems such as ext3 and FAT32 because it\n" \
-    "                              takes almost same time as 'prealloc' and it\n" \
-    "                              blocks aria2 entirely until allocation finishes.\n" \
+    "                              almost instantly. Note that btrfs disables\n" \
+    "                              compression when pre-allocating, use ``trunc``\n" \
+    "                              or ``none`` if you want compression. Don't use\n" \
+    "                              'falloc' with legacy file systems such as ext3\n" \
+    "                              and FAT32 because it takes almost same time as\n" \
+    "                              'prealloc' and it blocks aria2 entirely until\n" \
+    "                              allocation finishes.\n" \
     "                              'falloc' may not be available if your system\n" \
     "                              doesn't have posix_fallocate() function.\n" \
     "                              'trunc' uses ftruncate() system call or\n" \
